### DISCUSSION TITLE
Custom Error Page - not working

### DISCUSSION THREAD

**Discussion by GSBisran** [OrchardCMS/OrchardCore#7082] [Q&A]
I am using the latest version of orchard core cms and have defined all pages within the **Pages folder** as displayed below.
I am unable to find any documentation on how I can set up a custom error page to deal with **404 NotFound** and **internal 500** errors.

![directory structure](https://user-images.githubusercontent.com/71395434/93358006-e14d0680-f838-11ea-86ed-657bceeec900.png)

Would appreciate any help.
Thank you

**Comment by agriffard** [comment]
You can override the 404 by adding a NotFound.cshtml or .liquid file in Views/Shared:
https://github.com/OrchardCMS/OrchardCore/blob/dev/src/OrchardCore.Modules/OrchardCore.Diagnostics/Views/Diagnostics/NotFound.cshtml

https://github.com/OrchardCMS/OrchardCore/blob/dev/src/OrchardCore.Themes/TheBlogTheme/Views/Shared/NotFound.liquid

**Comment by orchardc** [comment]
That still doens't work.

I have put \Views\Shared\NotFound.cshtml

then browse to https://localhost:44360/test , still got the default asp.net not found error.

![image](https://user-images.githubusercontent.com/16193516/102484541-54476280-405e-11eb-88ad-c7f5893b939c.png)

I see there are reponse referencing to OrchardCore.Diagnostics, but how to make it work?? Maybe added to Nuget and then ?

Many thanks!

**Comment by jtkech** [comment]
@orchardc 

So you defined the not found page from the app, normally that's okay as an application behaves as a module and as a super theme.

We already set an higher precedence for the **shape template views** defined at the app level, but not for pure mvc views

Already fixed locally on my machine, i will do a PR soon, good catch.

**Reply by jtkech** [reply]
PR #8043 done ;)

**Comment by orchardc** [comment]
Thanks @jtkech , any idea how can I get this change into my code?

**Reply by jtkech** [reply]
You need to use the cloudsmith feed to get the last dev packages

**Reply by orchardc** [reply]
Hi @jtkech , I can't install the latest version 15648 with the following error.

Severity	Code	Description	Project	File	Line	Suppression State
Error	NU1102	Unable to find package OrchardCore.Setup.Abstractions with version (>= 1.0.0-rc2-15648)
  - Found 465 version(s) in Orchard Preview [ Nearest version: 1.0.0-rc2-15647 ]
  - Found 10 version(s) in nuget.org [ Nearest version: 1.0.0-rc2-13450 ]
  - Found 0 version(s) in Microsoft Visual Studio Offline Packages	

But I can install 15647 ; however the custom error still doesnt' work. Have you merged your change into 15647 or 15648 ?

**Reply by jtkech** [reply]
Yes, just saw on cloudsmith that there is a problem with this package for version 15648

But i think my change is included in 15647

Did you put your NotFound views under `Views/Shared`?

Try a rebuild

Check if under your bin folder you have a `YourAppName.Views.dll`

Show me your app project file

**Reply by jtkech** [reply]
Package issue fixed, you can now use 15650 or higher

**Comment by orchardc** [comment]
Hi @jtkech 
Under the bin folder I have the MyApp.views.dll

I also put the NotFound view under View/Shared folder, it also contains 1 line of code: 

`<h1>@T["Custom NotFound Page!"]</h1>`

![image](https://user-images.githubusercontent.com/16193516/102553149-4cb5a700-40ba-11eb-807f-1dd247f53ecf.png)

Is there anything else need to modify based on the out of box / standard MVC website project? 

Here's the startup.cs

```
 public class Startup
    {
        public Startup(IConfiguration configuration)
        {
            Configuration = configuration;
        }

        public IConfiguration Configuration { get; }

        public void ConfigureServices(IServiceCollection services)
        {
            services.AddOrchardCms();
        }

        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            app.UseHttpsRedirection();
            app.UseStaticFiles();
            app.UseStaticFiles();
            app.UseOrchardCore();
        }
    }
```

Here's the program.cs


```
    public class Program
    {
        public static void Main(string[] args)
        {
            CreateWebHostBuilder(args).Build().Run();
        }

        public static IWebHostBuilder CreateWebHostBuilder(string[] args) =>
            WebHost.CreateDefaultBuilder(args)
                .UseStartup<Startup>();

    }
```

**Comment by jtkech** [comment]
Looks good

I tried it with the full source code, tomorrow i will try it by using package references

Hmm, show me your `.csproj` file

Are the other views discovered?

**Reply by orchardc** [reply]
```
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <TargetFramework>netcoreapp3.1</TargetFramework>
    <PreserveCompilationReferences>true</PreserveCompilationReferences>
    <RuntimeIdentifier>win10-x64</RuntimeIdentifier>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.AspNetCore.Mvc.TagHelpers" Version="2.2.0" />
    <PackageReference Include="OrchardCore.Application.Cms.Core.Targets" Version="1.0.0-rc2-15650" />
    <PackageReference Include="OrchardCore.ContentManagement" Version="1.0.0-rc2-15650" />
    <PackageReference Include="OrchardCore.ContentManagement.Display" Version="1.0.0-rc2-15650" />
    <PackageReference Include="OrchardCore.Diagnostics" Version="1.0.0-rc2-15650" />
    <PackageReference Include="OrchardCore.DisplayManagement" Version="1.0.0-rc2-15650" />
    <PackageReference Include="OrchardCore.Menu" Version="1.0.0-rc2-15650" />
    <PackageReference Include="OrchardCore.ResourceManagement" Version="1.0.0-rc2-15650" />
    <PackageReference Include="OrchardCore.ResourceManagement.Abstractions" Version="1.0.0-rc2-15650" />
    <PackageReference Include="OrchardCore.Theme.Targets" Version="1.0.0-rc2-15650" />
    <PackageReference Include="OrchardCore.Setup.Abstractions" Version="1.0.0-rc2-15650" />
  </ItemGroup>
  <ItemGroup>
    <Folder Include="Repositories\" />
  </ItemGroup>

</Project>
```

**Reply by orchardc** [reply]
the other views only discoverable with a  [Route("/")] tag, for example for the **/home** to be used, I have to put the following code:

```
  [Route("/home")]
        public IActionResult Index()
        {
            return View();
        }
```

**Reply by orchardc** [reply]
The landing page is empty: 
![image](https://user-images.githubusercontent.com/16193516/102556539-0748a800-40c1-11eb-8db0-eef65da63da2.png)

Without the [Route] tag, it will only work if browse to: https://localhost:12345/MyApp/Home/Index

**Comment by orchardc** [comment]
Aha, the NotFound works propertly if I specify a Theme from here: /Admin/Themes

**Comment by jtkech** [comment]
Some remarks after a quick look

Don't use route in razor view, only if using razor page (see below)

Normally to render a "page" it is through one of our custom controller that display a content item that was created through the admin UI, and whose content type includes an `AutoroutePart` to define a route that can be marked as the `HomeRoute`. Then when we render a content item, all its parts / fields are rendered through shapes that a theme can override, these shapes are rendered through shape template views that are razor files but not pure mvc views

When we define a route in a razor file, usually we use razor pages (with a `@page` directive) this in a decoupled mode where the page use e.g. the content management services to get data, but not the display management services, so not the theming engine.

Not sure you need the `RuntimeIdentifier` property

Not related but you call twice `app.UseStaticFiles();`

Not good to reference `OrchardCore.Theme.Targets` from the main app

Anyway you just need to reference `OrchardCore.Application.Cms.Core.Targets`

Try also by referencing `OrchardCore.Application.Cms.Targets` in place of `Cms.Core` that doesn't include any front end theme.

Which setup recipe did you use on setup? I assume the blank site, right?

> Aha, the NotFound works propertly if I specify a Theme from here: /Admin/Themes

Just saw this comment, yes you need at least one selected theme, so that the theming engine is used and the app can override the current theme views

Hmm, but normally it should still work as the notfound is a mvc view, as i remember i tried it and it was working, hmm maybe because i had a selected front end theme just before, i will retry it

**Reply by jtkech** [reply]
I really suggest you to first use `OrchardCore.Application.Cms.Targets` and see how things work

This before using `OrchardCore.Application.Cms.Core.Targets` without any front end theme (only for the admin)

Then look at our source code, e.g. see the startup file of the `OC.Cms.Web` app

**Reply by orchardc** [reply]
Thanks for your suggestion; indeed I tried a lot. :)
Now I just keep the OrchardCore.Application.Cms.Targets, all works fine now.

**"Don't use route in razor view"**
However I can't find other way to specify the route. Based on the startup.cs and program.cs provided, could you suggest how to specify route in the startup.cs?

**Reply by jtkech** [reply]
So many way to define routes ;)

So usually we use the content item autoroute part through the admin UI, e.g. if you use the blog recipe on setup, then go to the admin in content definition to see how types / parts are defined, then see e.g. that the `Article` type contains an `AutoroutePart`, and so on.

Then yes for controllers you can use e.g. `MapAreaController` in the startup Configure of a given module, from the app don't do that directly, you would have to use our custom helpers so that there are registered as it is done through a module. We need this as ConfigureServices() / Configure() need to be done in a tenant context / container, but that's another story

Then yes you can decorate your controller with a route attribute to target mvc views (not shape template views), but doing this mean that you don't use our display management but a kind of decoupled mode where you grab data and you manage the rendering. In that case, in place of using razor views, we usually use razor `@page` that are also a kind of controllers.

**Reply by orchardc** [reply]
Hi @jtkech , the MapAraController has problem using the latest Orchard's Code Generation Templates `dotnet new ocmvc  `

**System.Exception: 'Could not resolve a service of type 'Microsoft.AspNetCore.Routing.IEndpointRouteBuilder' for the parameter 'routes' of method 'Configure' on type 'MySite2.Startup'.'**

startup.cs - I added IEndpointRouteBuilder 
```
    public class Startup
    {
        public void ConfigureServices(IServiceCollection services)
        {
            services.AddOrchardCms();
        }

        public void Configure(IApplicationBuilder app, IHostEnvironment env, IEndpointRouteBuilder routes)
        {
            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
            }

            app.UseStaticFiles();
            app.UseOrchardCore();

            routes.MapAreaControllerRoute(
               name: "Demo.Home.Index",
               areaName: "MySite2",
               pattern: "Home/Index",
               defaults: new { controller = "Home", action = "Index" }
            );

        }
    }
```


program.cs

```
    public class Program
    {
        public static Task Main(string[] args)
            => BuildHost(args).RunAsync();

        public static IHost BuildHost(string[] args)
            => Host.CreateDefaultBuilder(args)
                .ConfigureWebHostDefaults(webBuilder =>
                    webBuilder.UseStartup<Startup>())
                .Build();
    }
```

**Reply by jtkech** [reply]
When using OrchardCore most of the things are executed in a tenant container, we do this for multi tenancy and modularity, meaning that an application can serves multiple isolated tenants / sites, each one composed of different modules / features.

So, normally most of services registrations and pipeline configurations are done through the modules startups, not from the application startup, e.g. here you can't map routes as we didn't call `appBuilder.UseRouting().UseEndpoints(...)` from the app. But we do this for you for each tenant container (we have at least a `Default` tenant).

So, normally you would need to create modules / themes (that are specific modules), but as said, the app is also a module, so you can register / configure things as done by a regular module, but you need to use custom helpers on our `OrchardCoreBuilder` e.g. returned by `AddOrchardCms()`. They have different overloads / optional params, here an example

            services.AddOrchardCms()
                .ConfigureServices(tenantServices =>
                {
                    //tenantServices.AddScoped(...);
                })
                .Configure((tenantApp, tenantRoutes) =>
                {
                    // tenantRoutes.MapAreaControllerRoute(...)
                });

**Reply by orchardc** [reply]
Perfect, thanks a lot!!

**Reply by jtkech** [reply]
Cool, so mark it as answered if that's okay ;)
